### PR TITLE
Upgrade go version to 1.18

### LINKED ISSUE #5538: CVEs in latest distroless image due to 1.17.10 of go
REMOVED details of CVEs
Looks like upgrading to go 1.17.13 would fix these.  Is this a change that can be made?

### LINKED ISSUE #5561: Bump go version to 1.18 (or 1.19?)
For a security standpoint, we need to bump Go to a recent version.

### PR DESCRIPTION
**Pull request by aljesusg**
Closes #5561 
Fixes #5538

Upgrade go version 1.18.4

### CODE REVIEW

**Review by jshaughn — COMMENTED** [review]
Technically I don't think there would be an issue moving to Go 1.18, which should be backward compatible.   The question I have is only around compatibility with upstream console, and the plugin effort.  Also, to some degree, whether there is any build impact against older supported versions.

**Comment by jmazzitelli** [comment]
> The question I have is only around compatibility with ... the plugin effort

The [plugin](https://github.com/kiali/openshift-servicemesh-plugin/tree/main/plugin) doesn't use Go - strictly javascript.
The [plugin operator](https://github.com/kiali/openshift-servicemesh-plugin/tree/main/operator) doesn't use Go - strictly Ansible.

**Comment by lucasponce** [comment]
I wonder if:
- Keep the latest go 1.17 to be aligned with Istio 1.14.x https://github.com/istio/istio/blob/1.14.3/go.mod (if the issue is addressed there).
- Bump to go 1.18 when Kiali bumps to Istio 1.15 https://github.com/istio/istio/blob/master/go.mod#L3

In theory, it should work but using the same version of the dependencies https://github.com/istio/client-go/blob/v1.14.3/go.mod#L3 always helps to avoid potential issues.

But if the testing is solid, I imagine all would be good.

**Comment by aljesusg** [comment]
Mmm in this case I prefer to wait until istio 1.15 to be aligned.

What do you think?

**Comment by nrfox** [comment]
@aljesusg with istio 1.15 released can this be worked on again?

**Comment by jshaughn** [comment]
From Alberto:

> I wonder if:
Keep the latest go 1.17 to be aligned with Istio 1.14.x https://github.com/istio/istio/blob/1.14.3/go.mod (if the issue is addressed there).
Bump to go 1.18 when Kiali bumps to Istio 1.15 https://github.com/istio/istio/blob/master/go.mod#L3
In theory, it should work but using the same version of the dependencies https://github.com/istio/client-go/blob/v1.14.3/go.mod#L3 always helps to avoid potential issues.
But if the testing is solid, I imagine all would be good.
I don't know implications for 1.19 but again I need to check downstream @leandroberetta
Well this is a tricky question when upstream and downstream build with the same Go version. 1.57 is already targeted to 1.15 upstream. 1.59 will be targeted to 1.16 upstream. Yet 1.57 will be used with 1.14 downstream. We don't have to backport to previous versions, but we have already had requests to bump Go in the upstream.

I think we should proceed with updating to v1.18 for this sprint (1.59). We probably should backport the change to 1.58 becaue it will be the highest recommended version for use with 1.15. We do not have to backport to 1.57 because it will be superseded by 1.58 for the upstream use, and then it will remain at 1.17 for downstream 2.3/1.14.

**Comment by jshaughn** [comment]
@aljesusg Can we remove the Do Not Merge label?  I know the PR is not ready but it is active for this sprint, right?

**Comment by aljesusg** [comment]
Yes, I am checkin that all is running correctly

**Comment by aljesusg** [comment]
I disabled sa1029 while I check hwo to fix it.

handlers/metrics_test.go:348:57: SA1029: should not use built-in type string as key for value; define your own type to avoid collisions (staticcheck)
	req = req.WithContext(context.WithValue(req.Context(), "authInfo", &api.AuthInfo{Token: "test"}))


If I use a constant to do that or another way the tests fail I need to check the latest changes

File: graph/config/cytoscape/cytoscape.go
Comment (nrfox) [inline graph/config/cytoscape/cytoscape.go:260]: @aljesusg this could be the cause of the frontend failures from CI. You'll likely need to keep the assignment but maybe remove the type cast to `interface{}` to make the linter happy.

File: graph/config/cytoscape/cytoscape.go
Comment (nrfox) [inline graph/config/cytoscape/cytoscape.go:258]: nit: you don't need to add the linting rule here as a comment.

File: config/authentication.go
Comment (nrfox) [inline config/authentication.go:1]: I don't think this belongs in the `config` package. Can it be moved to `handlers`?

File: .github/workflows/build-backend.yml
Comment (jshaughn) [reply .github/workflows/build-backend.yml:26]: +1, this looks like a nicer approach.

File: Makefile
Comment (jshaughn) [inline Makefile:32]: I see there is now a 1.18.8, but I think we need to use what is available in Brew, right?

File: config/authentication.go
Comment (jshaughn) [reply config/authentication.go:1]: I think there is already an `authentication.go` in `handlers`.  Perhaps this can be merged in?   Or it can stay here if you think it makes more sense, but I agree that I don't currently see why this is config.

File: go.mod
Comment (jshaughn) [reply go.mod:28]: The prometheus updates make me a little nervous, just need to make sure the updated versions are compatible with the Prom versions we support for upstream and downstream.

File: graph/config/cytoscape/cytoscape.go
Comment (jshaughn) [inline graph/config/cytoscape/cytoscape.go:16]: Was this the linter or just personal preference?

File: graph/protocol.go
Comment (jshaughn) [inline graph/protocol.go:43]: same as above, I don't think we should be changing the comment style of the original author, unless there is some good reason.  Files are being updated in this PR just to change comment style or indentation.

**Review by jshaughn — COMMENTED** [review]
Very close, a couple of comments

File: .github/workflows/build-backend.yml
Comment (aljesusg) [reply .github/workflows/build-backend.yml:26]: Yes is read it from the go.mod

File: Makefile
Comment (aljesusg) [reply Makefile:32]: Yep in brew we have only until 1.18.7

File: config/authentication.go
Comment (aljesusg) [reply config/authentication.go:1]: Well the point is that we need to have this in the same file I don't care about where I thought in the config by the "authInfo" variable let me merge this in handlers

Comment (aljesusg) [reply config/authentication.go:1]: mmmm we'll have import cycle if merge this in handlers/authentication

File: graph/telemetry/istio/istio.go
Comment (aljesusg) [reply graph/telemetry/istio/istio.go:913]: You are right the warning was

S1009: should omit nil check; len() for github.com/prometheus/client_golang/api/prometheus/v1.Warnings is defined as zero (gosimple)

File: models/health.go
Comment (aljesusg) [reply models/health.go:75]: Yes... :(

File: prometheus/metrics.go
Comment (aljesusg) [reply prometheus/metrics.go:56]: Yes, the warning was S1009: should omit nil check; len() for github.com/prometheus/client_golang/api/prometheus/v1.Warnings is defined as zero (gosimple)

**Comment by aljesusg** [comment]
well maybe I did a go mod tidy .

Testing with:
go mod edit -go=1.18
 go get istio.io/client-go@v1.15.2

File: prometheus/metrics.go
Comment (jshaughn) [inline prometheus/metrics.go:99]:
```suggestion
	if len(warnings) > 0 {
```

**Review by jshaughn — CHANGES_REQUESTED** [review]
You need to be careful to catch all instances of the same change.  Please accept the suggested changes I added in the comments.  I suggested a different package for that code-move that I think would work and may make more sense that `config`.  See what you think.  I'm not sure about all of the comment changes, is that all due to the new linter, or is you IDE?  I just want to make sure that we don't get into an IDE formatting war.

File: prometheus/metrics.go
Comment (jshaughn) [inline prometheus/metrics.go:179]:
```suggestion
		log.Warningf("getRequestRatesForLabel. Prometheus Warnings: [%s]", strings.Join(warnings, ","))
```

**Review by jshaughn — CHANGES_REQUESTED** [review]
Almost there.  The CI is now failing.  I suggested one change for a small issue in an error message, not something you actually changed.  Were you able to ignore those jaeger pb.go files from the linter (I'mreferring to an older comment from Nick)?

**Review by jshaughn — APPROVED** [review]
Code changes look good to me.  You didn't really answer my questions about the changes to comments, as to whether they were required by the new linter or just things you wanted to change.  I'll assume they were forced by the linter.  Approved!

**Comment by aljesusg** [comment]
Yes sorry . Of course I hate spend time change unnecessary code like comments 😅😅 I tried gofmt and others and linter throw a lot of warnings due spaces, tabs, indentation 😔

**Review by nrfox — APPROVED** [review]
LGTM

**Pull request merged**
